test/connectors/test_torch_connector.py

```diff
@@ -659,6 +659,205 @@ def test_circuit_qnn_batch_gradients(self, config):
             places=4,
         )
 
+    @requires_extra_library
+    def test_opflow_qnn_hybrid_batch_gradients(self):
+        """Test gradient back-prop for batch input in hybrid opflow qnn."""
+
+        try:
+            from torch import cat
+            from torch.nn import Linear, Module
+            import torch.nn.functional as F
+        except ImportError as ex:
+            raise MissingOptionalLibraryError(
+                libname="Pytorch",
+                name="TorchConnector",
+                pip_install="pip install 'qiskit-machine-learning[torch]'",
+            ) from ex
+
+        num_inputs = 2
+
+        # set up QNN
+        qnn = TwoLayerQNN(
+            num_qubits=num_inputs,
+            quantum_instance=self.sv_quantum_instance,
+            input_gradients=True,  # for hybrid qnn
+        )
+
+        # set up dummy hybrid PyTorch module
+        class Net(Module):
+            """Pytorch nn module."""
+
+            def __init__(self):
+                super().__init__()
+                self.fc1 = Linear(4, 2)
+                self.qnn = TorchConnector(
+                    qnn, np.array([0.1] * qnn.num_weights)
+                )  # Apply torch connector
+                self.fc2 = Linear(1, 1)  # 1-dimensional output from QNN
+
+            def forward(self, x):
+                """Forward pass."""
+                x = F.relu(self.fc1(x))
+                x = self.qnn(x)  # apply QNN
+                x = self.fc2(x)
+                return cat((x, 1 - x), -1)
+
+        model = Net()
+
+        # random data set
+        x = Tensor(np.random.rand(5, 4))
+        y = Tensor(np.random.rand(5, 2))
+
+        # define optimizer and loss
+        optimizer = SGD(model.parameters(), lr=0.1)
+        f_loss = MSELoss(reduction="sum")
+
+        # loss and gradients without batch
+        optimizer.zero_grad(set_to_none=True)
+        sum_of_individual_losses = 0.0
+        for x_i, y_i in zip(x, y):
+            output = model(x_i)
+            sum_of_individual_losses += f_loss(output, y_i)
+        sum_of_individual_losses.backward()
+        sum_of_individual_gradients = 0.0
+        for n, param in model.named_parameters():
+            name = n.split(".")
+            # make sure gradient is not None
+            self.assertFalse(param.grad is None)
+            if name[1] == "weight" or name[1] == "_weights":
+                sum_of_individual_gradients += np.sum(param.grad.numpy())
```

Comment:
Can you please explain what's going on here and why we need this `if` statement? Please ping me to discuss this part when you have time.

Reply:
My comment

>Having to rely on a private variable, _weights, whatever the name, is not good,

Was in direct relation to this code seeing the name being used here. Hence the question was this what we would normally expect an end user to do.

> Anyway, the parameter values are managed automatically by Torch's differentiation engine (it just detects the TorchParameter class and takes care of everything parameter-related), so can they be considered private?

All variables are accessible in Python whether they are considered private or not. The main difference is that we only consider the public API when doing changes and deprecate if/when compatibility breaks. As part of the private interface we freely change that, since its supposed to be part of the private implementation. Hence my concern if that access was considered how a user would normally do things I think we need to look at that!

Reply:
We can avoid having both `_weights` and `weight`. I propose to introduce module parameters explicitly:
```python
        weight = TorchParam(Tensor(neural_network.num_weights))
        self.register_parameter("weight", weight)
        self._weights = weight    # this is for compatibility only, can be revised.
```
This creates a parameter on a module level that can be accessed in the same fashion as we access traditional layers. Thus, we can simplify the `if` statement and remove our property `weight`. In the same time we can keep `self._weights` for compatibility.

Or we can simply rename `_weights` in favor of `weight` to be compliant with pytorch but incompliant with the rest code. @stefan-woerner any opinion?

Reply:
@adekusar-drl True! Registering the parameter with a different name is a good idea. I don't know why I didn't think of that before... I will wait for consensus and make the corresponding changes.

Reply:
@ElePT Can you verify that registering parameter explicitly does not break anything? If this is the one things we should do to have compatible parameter naming, then, I think, we can proceed with this option.

Reply:
@adekusar-drl It works as expected, nothing broken as far as I can tell, so I have already pushed the changes. As this affects the way the user interacts with the TorchConnector, should I also add a release note?

Reply:
I'm not sure it affect the way a user interacts with TorchConnector. There's the same property, I don't see how the changes may affect users, so in this case it is not required. Well, if I don't miss anything.

Reply:
I mean... it changes the output of  the torch function `model.named_parameters()`. Before , it would output `{'qnn._weights', 2.5}`, and now it will output `{'qnn.weight', 2.5}` (just like in the unit test). If somebody wants to check the gradients, they will see that difference in the name (that's why we changed it). It's true that it's not a direct interaction with the user, but maybe an interaction through PyTorch? I am asking because I it's not that clear to me either.

Reply:
Now I see your point. Yes, better to have this change documented in a reno. Thanks for noting this.